### PR TITLE
[CR] Operation: AIM high!

### PR DESCRIPTION
**Pull request by DavidKeaton**
#### This should be ready to merge, as the unchecked entities will be added later! :-)

Thanks again to @Coolthulhu for being _the man_ in helping me get some of this stuff fixed and merged while I was away (I appreciate it duder).
##### This will address the following: (this list may or may not grow)
- [x] "It is still not possible to move all items under a vehicle cargo. They'll be placed in the cargo or not placed at all."
- [x] "It is not possible to pick vehicle tile and the tile under this one at the same time."
- [ ] "When targeting worn items and moving 2 (or more) items of the same type, only 1 of them will be worn."
- [ ] Source area to work with `AIM_ALL`, so you can move everything into a neat pile.
- [x] Fix containers and all its containing powers.
- [x] Fix `maximum-1` on moving stacks (revert to old usage).
- [x] Enable moving bulk items from anywhere -> inventory (works already with inventory -> anywhere).
- [ ] Fix filter string parsing when first entry is not a negation. (#12110)
- [x] `move_all_items()` to work for different pane settings.
- [x] `move_all_items()` with stacked items (#12205)
- [x] Update minimap to display new/correct information regarding new AIM abilities.
- [x] center tile rudeness (#12212)
- [x] `AIM_DRAGGED` duplication in `AIM_ALL` (#12207)

-Davek

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Wait until #12137 is merged. It does some of the things from your list and you'll certainly run into conflicts if it is merged when you're halfway done.
My PR fixes two first points from your list.

Or you can merge it locally and work on a branch that includes the changes. To do that, you'd only need to invoke `git pull https://github.com/Coolthulhu/Cataclysm-DDA.git aim-vehicle-select` while in your current branch.
Then your PR would not gain any conflicts when mine gets merged.

Some minor things that could but don't need to be a part of this PR:
* "Fix maximum-1 on moving stacks (revert to old usage)." - this should be the case when you're forced to only move a part of the stack (due to the entire stack being too big), but if possible, manual invocation by player who wants to move only a part of the stack should not pre-input any number.
* Kevin wants AIM to eventually be incapable of reaching through vehicle floor. Currently it can do so because 'e'xamine can do it to, but eventually neither should be capable of it.
* graysage1 brought up some interesting points on typical AIM usage: he starts by resetting it with 3 keypresses. For example, to move from left to right, you can press right,right,left. This is quite convenient. Maybe AIM could be reworked to always change the active pane on a direction change? That way you could always set it up in just two keypresses: first source, then destination.
* Resetting AIM state every time was quite annoying, even during testing. AIM should either remember everything or give an option to be reset, not unconditionally reset stuff.
* Technical: it would be useful if locations or squares carried info about vehicles and not just potential vehicles but targetted one. Currently the only way to check for vehicle target is to check if the vehicle exists and if the current pane targets it, which involves two different classes. Not mandatory or anything, but it could make further work on the code easier.

**Comment by DavidKeaton** [comment]
right on, I'll work from your branch. is it simply waiting for merge, or still worked on?

**Comment by Coolthulhu** [comment]
Waiting for merge, unless someone finds a bug in it.

**Comment by DavidKeaton** [comment]
ah, don't want to pull a Davek? ;-)

**Comment by Rivet-the-Zombie** [comment]
>Kevin wants AIM to eventually be incapable of reaching through vehicle floor. Currently it can do so because 'e'xamine can do it to, but eventually neither should be capable of it.

This should really require a system whereby enemies killed on vehicle tiles don't fall through to the ground below. Otherwise, if you kill a zed inside a car their clothes and corpse will end up somehow teleporting to the inaccessible space beneath the vehicle.

**Comment by graysage1** [comment]
@Coolthulhu 's list contains almost everything, but can we also consider fixing the bug mentioned in  #12110?  I would appreciate it.  It's not as high priority (for me at least) as the stuff already on your list, but unless you can't dupe this problem, technically it's still a bug. ;)

**Comment by graysage1** [comment]
@Rivet-the-Zombie 

>This should really require a system whereby enemies killed on vehicle tiles don't fall through to the ground below. Otherwise, if you kill a zed inside a car their clothes and corpse will end up somehow teleporting to the inaccessible space beneath the vehicle.

I think you make a good point, and besides that, if the car tile with the dying Zed is already full, the stuff may need to spill over to the next tile, which itself can be full, and so on.

If I understand correctly, car tiles rarely can contain 4000 volume units they way ground tiles can.  So on the ground spillover to the next tile is rarely, if ever needed.  But on a vehicle it may be needed.  If I am right, this will complicate things even further.

**Comment by BevapDin** [comment]
> This should really require a system whereby enemies killed on vehicle tiles don't fall through to the ground below.

That is just one example.

Players *need* to be able to access items under any vehicle tile. This is simply a result of many other places in the code that put stuff directly on the map, e.g. monster death drops (including the corpse, which will often (fat zombie) not fit into *any* vehicle storage anyway), crafting/uncrafting products, smashing vehicle parts byproducts (and cargo of destroyed cargo parts etc.), etc. search for `m.add_item` or `m.spawn_item`, they all put things directly on the map.

There is also this scenario:
```
+++
+@+
+++
```
The character is inside a vehicle on top of an aisle (no storage), surrounded by obstacles (closed doors, those without storage, or windshields). Now craft, or drop or whatever, where do the items go? If they go on the map, why can't the character access them? If they vanish, they ... I'm not even considering this.

Similar, make the aisle two tiles width and imprison two creatures (hostile to each other), one of them dies, where does the corpse go?

Similar: all the same, but add cargo capacity to all vehicle tiles, make the cargo parts full. Now, where do the items go?

This is btw. already a problem with the sealed furniture. If I plant seeds on a 10x10 area (that is 100 tiles, each having a f_seedling furniture), stand in the middle and drop an item. Where does it vanish to?

**Comment by DavidKeaton** [comment]
@graysage1 Yeah man, added to the list. :-)

**Comment by kevingranade** [comment]
> Examples of item handling contradictions due to enclosed spaces that can't carry items

This isn't vehicle specific, do the same thing in a closet with the door closed and you have the same issue.

I much prefer the somewhat-silly solution of having items migrate to the edge of the vehicle and fall out to the current situation where a player can routinely reach through the floor of a car to pick up items on the ground.

**Comment by DavidKeaton** [comment]
@BevapDin
Interesting notion at the end there. Where does it vanish to? Does it keep crawling outwards until it reaches a tile where the item _can_ be stored?

**Comment by Coolthulhu** [comment]
@DavidKeaton Will fail when it can't find a good spot within radius.

**Comment by DavidKeaton** [comment]
interesting, but good. makes no sense to tidal wave it. also @Coolthulu, pulled your branch in, fixing up minor merge conflicts now.

**Comment by DavidKeaton** [comment]
gonna push in all my changes soon

**Comment by Rivet-the-Zombie** [comment]
Any progress on this?

**Comment by DavidKeaton** [comment]
yes, adding fixes here and there.

**Comment by DavidKeaton** [comment]
actually @Rivet-the-Zombie. where it is ATM may be enough of the fixes to be merge-able. the other stuff is minor and can go in afterwards. the bugs should be gone now.

File: src/uistate.h [hunk 1]
```diff
@@ -8,46 +11,52 @@
...
-        int  adv_inv_veh_location_dest;
-
-        bool editmap_nsa_viewmode;        // true: ignore LOS and lighting
-        bool overmap_blinking;            // toggles active blinking of overlays.
-        bool overmap_show_overlays;       // whether overlays are shown or not.
+        int wishitem_selected = 0;
+        int wishmutate_selected = 0;
+        int wishmonster_selected = 0;
+        int iexamine_atm_selected = 0;
+        std::vector<int> adv_inv_sort = {1, 1};
```
Comment (BevapDin) [inline src/uistate.h]: This (and the other vectors) should probably be `std::array<int, 2>`. I don't expect it to store an arbitrary number of elements, but *always* 2, so why allow resizing it?.

File: src/enums.h [hunk 2]
```diff
@@ -290,7 +290,27 @@ struct tripoint : public JsonSerializer, public JsonDeserializer {
     {
         return tripoint( -x, -y, -z );
     }
-
+    /*** some point operators and functions ***/
+    tripoint operator+(const point &rhs) const
+    {
```
Comment (BevapDin) [inline src/enums.h]: This operator is already implementation (see below), but it is not const. Please remove this implementation and add the const to the other one.

File: src/advanced_inv.cpp [hunk 3]
```diff
@@ -99,6 +91,38 @@ advanced_inventory::~advanced_inventory()
...
+}
+
+void advanced_inventory::load_settings()
+{
+    bool moved = uistate.adv_inv_last_coords != g->u.pos();
+    static const std::array<aim_location, NUM_PANES> default_areas = {
+        {AIM_ALL, AIM_INVENTORY}
+    };
+    for(int i = 0; i < NUM_PANES; ++i) {
+        auto location = (moved == true) ? default_areas[i] :
```
Comment (BevapDin) [inline src/advanced_inv.cpp]: This (specifically the `moved == true` part) ignores the DRAGGED vehicle setting. If the character has moved, the default area is used (btw. a clever solution, the array). This makes that location basically useless.

You don't like the boolean negation `!` at all, do you? (-:

File: src/activity_item_handling.cpp [hunk 4]
```diff
@@ -361,13 +380,14 @@ static void move_items( const tripoint &source_arg, const tripoint &destination_
 
             // Drop it first since we're going to delete the original.
             dropped_items.push_back( *temp_item );
-            g->drop( dropped_items, dropped_worn, 0, destination.x, destination.y );
+            // no 3d function yet
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: Why a re you calling it with a tripoint as argument when it's not a 3d function?

File: src/advanced_inv.h [hunk 5]
```diff
@@ -336,7 +351,13 @@ class advanced_inventory
 
         bool exit;
 
+        // store/load settings (such as index, filter, etc)
+        void save_settings(bool only_panes = false);
```
Comment (BevapDin) [inline src/advanced_inv.h]: The function is called at two places, an explicit value for the parameter is given at one of them. Why don't you add the default value to the other place (in the destructor) and make the parameter mandatory. That actually saves you 3 characters: ` = ` (-:

File: src/advanced_inv.h [hunk 6]
```diff
@@ -336,7 +351,13 @@ class advanced_inventory
...
+        void load_settings();
+        void do_return_entry();
+
         static std::string get_sortname(advanced_inv_sortby sortby);
+        // if called with default arg, will exit after first completion (if not AIM_ALL)
```
Comment (BevapDin) [inline src/advanced_inv.h]: I don't get the comment. The function has no parameters at all. There is no default argument either.

File: src/advanced_inv.cpp (hunk 3 above)
Comment (DavidKeaton) [inline src/advanced_inv.cpp]: I'm not against '!', but it seems I've been told clarity is more important. ;-) I use AIM the same way, so i'll look at this. :-)

File: src/advanced_inv.h (hunk 6 above)
Comment (DavidKeaton) [inline src/advanced_inv.h]: may have been left over, my bad. :-)

File: src/uistate.h (hunk 1 above)
Comment (DavidKeaton) [inline src/uistate.h]: never thought of that. :-) hehe.

File: src/enums.h (hunk 2 above)
Comment (DavidKeaton) [inline src/enums.h]: can do

**Comment by Zireael07** [comment]
How close are those fixes to getting merged?

**Comment by DavidKeaton** [comment]
sans some minor fixes BevapDin pointed out, it should address all that has been checked. it is p much good to go. :-)

**Comment by BevapDin** [comment]
(in the sorting of the items, using the sort-by-category order)
```C++
    // FIXME: TESTME!!!
    if(d1.idx != d2.idx) {
        return d1.idx < d2.idx;
    }
```
What is that supposed to do? `idx` is the index of the item (in the source item stack), using it here sorts the items by their position in the source stack. The category sort is supposed to sort by category first and by *item name* second. This will just sort by the order the items were added, this is not what it is supposed to be. If you want a new ordering (sorted by category first, by original ordering secondly), that's fine, just do it in a separate PR.

----

Maybe you should commit the changes as separate PRs? One could be the move to 3D points, another the changes to the `uistatedata`, another the to_vehicle / from_vehicle changes. It would be fine if you only fix one bug in each PR.

However, if you fix the sorting, this shall be merged.

----

Not a deal breaker, but annoying: If the code is not used, remove it, don't comment it out. The source control system has the old code anyway. If it's needed again, one can look it up there. As comment it only litters the code and nobody knows why it's still there (is it work in progress? comment out because it's not used? ...)

----

And I have a new AIM bug (most likely unrelated to the changes here, needs the option "Close AIM on move all" to be false): Open AIM, select inventory and some other place with several items, use "move all items" to move them all into the inventory. The AIM should have opened again, displaying a full inventory. Now choose "move all items" again to move the items from the inventory back to some place on the map. It will ask you whether you're sure, say yes. Nothing happens. Selecting another destination won't fix it either. Leaving the AIM and coming back (without anything between, just press ESC to leave and '/' to re-enter AIM) will fix it.

The cause seems to be that the AIM is opened from `player_activity::do_turn` when it has moved all items into the inventory. Selecting "move all" there will build up a player activity, but after leaving the AIM, `player_activity::do_turn` continues to call `player_activity::finish`, which will (the default case is used) set the activity type to ACT_NULL and clear the values vector, effectively canceling the activity.

Leaving the AIM and coming back means the AIM is entered from game.cpp and (upon leaving it) `player_activity::finish` is *not* called, the new activity remains.

File: src/advanced_inv.cpp [hunk 7]
```diff
@@ -861,7 +840,8 @@ void advanced_inventory_pane::add_items_from_area( advanced_inv_area &square,
             if( !cont->is_container_empty() ) {
                 // filtering does not make sense for liquid in container
                 item *it = &( square.get_container( in_vehicle() )->contents[0] );
-                advanced_inv_listitem ait( it, 0, 1, square.id );
+                // FIXME: TESTME for FALSE ???
+                advanced_inv_listitem ait( it, 0, 1, square.id, false );
```
Comment (BevapDin) [inline src/advanced_inv.cpp]: Uneducated guess, but why not use the value from `in_vehicle()`, as it is apparently used to get the container item in the first place? It probably doesn't even matter at all as the item (which is the contained liquid) is referenced through it's container and that is retrieved through the pane (not through the `advanced_inv_listitem`).

**Comment by DavidKeaton** [comment]
I will look into these fixes, thanks for taking the time to take a peek @BevapDin :-)

**Comment by BevapDin** [comment]
> Edit: by fixing the sorting, you mean to remove the little bit I added, correct? :-)

Yes, I don't see the reason for that change. If you want a  new ordering, just PR it, I see nor problem with that.

**Comment by DavidKeaton** [comment]
noted, thanks friend. :-)

**Pull request merged**